### PR TITLE
HHH-18588 fix minor defects in Hibernate Introduction article

### PR DESCRIPTION
**Pull request by NathanQingyangXu**
https://hibernate.atlassian.net/browse/HHH-18588

Fixed some minor defects.

### CODE REVIEW

File: documentation/src/main/asciidoc/introduction/Introduction.adoc [hunk 1]
```diff
@@ -13,7 +13,7 @@ Hibernate makes *relational data* visible to a program written in Java, in a *na
 3. allowing performance optimizations to be made after the basic persistence logic has already been written.
 ****
 
-Here the relational data is the focus, along with the importance of typesafety.
+Here the relational data is the focus, along with the importance of type safety.
```
Comment (NathanQingyangXu) [inline documentation/src/main/asciidoc/introduction/Introduction.adoc:16]: not sure the change is valid, but "type safety" was used later in the same chapter.

Comment (gavinking) [reply documentation/src/main/asciidoc/introduction/Introduction.adoc:16]: yeah I go back and forth on this. I like "typesafety" as a word, but I accept that "type safety" is more standard.

File: documentation/src/main/asciidoc/introduction/Introduction.adoc [hunk 2]
```diff
@@ -111,7 +111,7 @@ Most people implement the domain model as a set of what we used to call "Plain O
 
 [TIP]
 ====
-Take your time with this code, and try to produce a Java model that's as close as reasonable to the relational data model. Avoid using exotic or advanced mapping features when they're not really needed.
+Take your time with this part, and try to produce a Java model that's as close as reasonable to the relational data model. Avoid using exotic or advanced mapping features when they're not really needed.
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Introduction.adoc:114]: not sure why you propose this change.

File: documentation/src/main/asciidoc/introduction/Introduction.adoc [hunk 3]
```diff
@@ -542,7 +542,7 @@ public List<Book> findBooks(String titlePattern) {
 }
 ----
 
-You might be thinking that our query method looks a bit boilerplatey.
+You might be thinking that our query method looks a bit boilerplate.
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Introduction.adoc:545]: This was intentional. I'm allowed to invent words like "boilerplatey" because they're run and I like saying them :)

File: documentation/src/main/asciidoc/introduction/Introduction.adoc [hunk 4]
```diff
@@ -614,15 +614,15 @@ We do need to be careful here if our persistence code uses native SQL, or if it
...
+We _usually_ do this when we create the Hibernate `SessionFactory` or JPA `EntityManagerFactory`, and so traditionally we've used a <<automatic-schema-export,configuration property>> for this.
 
 The JPA-standard property is `jakarta.persistence.schema-generation.database.action`.
 For example, if we're using `Configuration` to configure Hibernate, we could write:
 
 [source,java]
 ----
 configuration.setProperty(AvailableSettings.JAKARTA_HBM2DDL_DATABASE_ACTION,
-                          Action.SPEC_ACTION_DROP_AND_CREATE);
+                          Action.CREATE);
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Introduction.adoc:625]: This is `org.hibernate.tool.schema.Action.SPEC_ACTION_DROP_AND_CREATE` I'm talking about. So I think this is correct.

File: documentation/src/main/asciidoc/introduction/Introduction.adoc [hunk 5]
```diff
@@ -699,7 +699,7 @@ This is again the job of the schema management tooling, either:
 [source,java]
 ----
 configuration.setProperty(AvailableSettings.JAKARTA_HBM2DDL_DATABASE_ACTION,
-                          Action.ACTION_VALIDATE);
+                          Action.VALIDATE);
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Introduction.adoc:702]: Likewise, I think this is correct.

File: documentation/src/main/asciidoc/introduction/Introduction.adoc (hunk 2 above)
Comment (NathanQingyangXu) [reply documentation/src/main/asciidoc/introduction/Introduction.adoc:114]: When I saw "this code", my first reaction is some concrete code snippet, so it confused me. I have no objection to revert it back for sure.

File: documentation/src/main/asciidoc/introduction/Introduction.adoc (hunk 3 above)
Comment (NathanQingyangXu) [reply documentation/src/main/asciidoc/introduction/Introduction.adoc:545]: ok, I googled a lot and no wonder I didn't find it anywhere, :).

Comment (NathanQingyangXu) [reply documentation/src/main/asciidoc/introduction/Introduction.adoc]: My original thought is if it is a new adjective word, it might be "boilerplaty" instead of "boilerplatey"? (inasmuch as 'taste' vs 'tasty'), but maybe you intend some fun here. Anyway, I revert back this change, among others.

File: documentation/src/main/asciidoc/introduction/Configuration.adoc [hunk 6]
```diff
@@ -2,10 +2,11 @@
 == Configuration and bootstrap
 
 We would love to make this section short.
-Unfortunately, there's several distinct ways to configure and bootstrap Hibernate, and we're going to have to describe at least two of them in detail.
+Unfortunately, there are several distinct ways to configure and bootstrap Hibernate, and we're going to have to describe at least two of them in detail.
 
 The four basic ways to obtain an instance of Hibernate are shown in the following table:
 
+.Four ways to bootstrap Hibernate
```
Comment (NathanQingyangXu) [inline documentation/src/main/asciidoc/introduction/Configuration.adoc:9]: without this table title, the next table will have title of table 2, a little bit inconsistent.

File: documentation/src/main/asciidoc/introduction/Configuration.adoc [hunk 7]
```diff
@@ -2,10 +2,11 @@
...
 [%breakable,cols="50,50"]
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Configuration.adoc:10]:
```suggestion
[%breakable,cols="50,50",number=0]
```

File: documentation/src/main/asciidoc/introduction/Configuration.adoc [hunk 8]
```diff
@@ -175,7 +176,7 @@ Each `<class>` element specifies the fully-qualified name of an entity class.
 
 .Scanning for entity classes
 ****
-In some container environments, for example, in any EE container, the `<class>` elements are unnecessary, since the container will scan the archive for annotated classes, and automatically recognize any class annotated `@Entity`.
+In some container environments, for example, in any EE container, the `<class>` elements are unnecessary, since the container will scan the archive for annotated classes, and automatically recognize any class annotated with `@Entity`.
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Configuration.adoc:179]: I prefer just "annotated `@Entity`". 

"With" is unnecessary here and sounds unnatural to me. I would never use "with" here in regular speech.

**Review by gavinking — CHANGES_REQUESTED** [review]

**Comment by NathanQingyangXu** [comment]
@gavinking , I finished checking the whole doc. I learned a lot during the process. Thanks for the brillant and insightful documentation!

**Comment by NathanQingyangXu** [comment]
I also created a similar PR for Hibernate Query Language doc at https://github.com/hibernate/hibernate-orm/pull/8941

File: documentation/src/main/asciidoc/introduction/Entities.adoc [hunk 9]
```diff
@@ -649,7 +649,7 @@ There are two ways to apply a converter:
 - the `@Convert` annotation applies an `AttributeConverter` to a particular entity attribute, or
 - the `@Converter` annotation (or, alternatively, the `@ConverterRegistration` annotation) registers an `AttributeConverter` for automatic application to all attributes of a given type.
 
-For example, the following converter will be automatically applied to any attribute of type `BitSet`, and takes care of persisting the `BitSet` to a column of type `varbinary`:
+For example, the following converter will be automatically applied to any attribute of type `EnumSet<DayOfWeek>`, and takes care of persisting the `EnumSet<DayOfWeek>` to a column of type `INTEGER`:
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Entities.adoc:653]: What was wrong with using `BitSet` in this example?

File: documentation/src/main/asciidoc/introduction/Entities.adoc [hunk 10]
```diff
@@ -1233,7 +1233,7 @@ class Book {
 
 Remember, if we wish to the modify the collection we must <<bidirectional-problem,change the owning side>>.
 
-We've again used ``Set``s to represent the association.
+We've again used `Set` to represent the association.
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Entities.adoc:1236]: Asciidoc requires double-backticks when there are no spaces around the literal. So double backticks here are necessary.

File: documentation/src/main/asciidoc/introduction/Entities.adoc [hunk 11]
```diff
@@ -1486,7 +1486,7 @@ Let's pause to remember the annotations we've met so far.
 | Annotation | Purpose | JPA-standard
 
 | `@GeneratedValue` | Specify that an identifier is system-generated | &#10004;
-| `@SequenceGenerator` | Define an id generated backed by on a database sequence | &#10004;
+| `@SequenceGenerator` | Define an id generated backed by a database sequence | &#10004;
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Entities.adoc:1489]:
```suggestion
| `@SequenceGenerator` | Define an id generator backed by a database sequence | &#10004;
```

File: documentation/src/main/asciidoc/introduction/Interacting.adoc [hunk 12]
```diff
@@ -995,7 +997,7 @@ The `getResultCount()` method is useful for displaying the number of pages of re
 SelectionQuery<Book> query =
         session.createSelectionQuery("from Book where title like ?1 order by title", Book.class)
             .setParameter(1, titlePattern);
-long pages = query.getResultCount() / MAX_RESULTS;
+long pages = (long) Math.ceil(query.getResultCount() * 1.0 / MAX_RESULTS);
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Interacting.adoc:1000]: Why is this change necessary?

**Review by gavinking — CHANGES_REQUESTED** [review]
@NathanQingyangXu please see my comments

File: documentation/src/main/asciidoc/introduction/Entities.adoc (hunk 9 above)
Comment (NathanQingyangXu) [reply documentation/src/main/asciidoc/introduction/Entities.adoc:653]: Screenshot 2024-10-03 at 08 40 00
Here is the context. It is confusing, for the Converter example just below is not using `BitSet` but `EnumSet<DayOfWeek>`. Then after this example class, `BitSet` is used again. 
I am trying to make both consistent. Using `EnumSet` instead of `BitSet` could avoid rewriting that converter class.

File: documentation/src/main/asciidoc/introduction/Interacting.adoc (hunk 12 above)
Comment (NathanQingyangXu) [reply documentation/src/main/asciidoc/introduction/Interacting.adoc:1000]: suppose the total number is 50 and max result per page is 20, then the page number should be 3 (`Math.ceil(50.0 / 20)`, the last page has 10 results) instead of 2 (`50 / 20` in Java).

Not sure my implementation is best though.

Comment (gavinking) [reply documentation/src/main/asciidoc/introduction/Interacting.adoc:1000]: I'm just not sure I wanna be blinding the reader with that level of detail here.

File: documentation/src/main/asciidoc/introduction/Entities.adoc (hunk 9 above)
Comment (gavinking) [reply documentation/src/main/asciidoc/introduction/Entities.adoc:653]: Oh, yeah, sure, you're right.

File: documentation/src/main/asciidoc/introduction/Entities.adoc (hunk 10 above)
Comment (gavinking) [reply documentation/src/main/asciidoc/introduction/Entities.adoc:1236]: But it's pluralized because we used `Set` twice, once on each side of the association, in the preceding code examples.

I don't see anything wrong here, and I don't think the change is an improvement.

File: documentation/src/main/asciidoc/introduction/Interacting.adoc (hunk 12 above)
Comment (gavinking) [reply documentation/src/main/asciidoc/introduction/Interacting.adoc:1000]: Perhaps the following is more immediately readable:

```java
long results = query.getResultCount();
long pages = results / MAX_RESULTS + (results % MAX_RESULTS == 0 ? 0 : 1);
```

**Comment by NathanQingyangXu** [comment]
Yeah, less tricky and mysterious. Will update as suggested later in my off
hours. It is a good catch though.

File: documentation/src/main/asciidoc/introduction/Entities.adoc (hunk 10 above)
Comment (NathanQingyangXu) [reply documentation/src/main/asciidoc/introduction/Entities.adoc]: I reverted back.

File: documentation/src/main/asciidoc/introduction/Interacting.adoc (hunk 12 above)
Comment (NathanQingyangXu) [reply documentation/src/main/asciidoc/introduction/Interacting.adoc]: changed as suggested

File: documentation/src/main/asciidoc/introduction/Interacting.adoc [hunk 13]
```diff
@@ -1061,7 +1064,7 @@ if (!firstPage.isLastPage()) {
 ----
 
 The "key" in key-based pagination refers to a unique key of the result set which determines a total order on the query results.
-In this example, `Book.isbn` is the key.
+In this example, `Book_.isbn` is the key.
```
Comment (gavinking) [inline documentation/src/main/asciidoc/introduction/Interacting.adoc:1067]: @NathanQingyangXu please revert this one. The key is the `isbn` field of `Book`, not its representation in the metamodel.

**Review by gavinking — CHANGES_REQUESTED** [review]

File: documentation/src/main/asciidoc/introduction/Interacting.adoc (hunk 13 above)
Comment (NathanQingyangXu) [reply documentation/src/main/asciidoc/introduction/Interacting.adoc]: reverted back

**Review by gavinking — APPROVED** [review]

**Comment by gavinking** [comment]
Thanks, @NathanQingyangXu !

**Comment by gavinking** [comment]
Ugh, I should have squashed this. I did not notice it was spread over multiple commits. Oops.

**Comment by NathanQingyangXu** [comment]
You might well revert commits already pushed. E.g. as per
https://gist.github.com/gunjanpatel/18f9e4d1eb609597c50c2118e416e6a6
